File: api/src/main/java/xyz/outlinr/api/service/EscrowService.java
package xyz.outlinr.api.service;

import xyz.outlinr.api.dto.request.*;
import xyz.outlinr.api.dto.response.*;
import xyz.outlinr.api.entity.User;
import xyz.outlinr.api.entity.enumeration.EscrowStatus;

import java.util.List;
import java.util.UUID;

public interface EscrowService {

    EscrowResponse createEscrow(CreateEscrowRequest request, User creator);

    EscrowResponse createAndFundEscrow(CreateEscrowRequest request, User creator, String txnRef, long amountInKobo);

    EscrowResponse getEscrow(UUID id, User currentUser);

    List<EscrowResponse> listEscrows(User currentUser, EscrowStatus statusFilter);

    EscrowResponse transitionStatus(UUID id, StatusTransitionRequest request, User currentUser);

    AuthResult acceptInvite(InviteAcceptRequest request);

    InvitePreviewResponse getInvitePreview(UUID inviteToken);

    void deleteEscrow(UUID id, DeleteEscrowRequest request, User currentUser);
}

File: app/src/services/escrow-service.ts
import { api } from "./api-client";
import type {
    CreateEscrowRequest,
    EscrowResponse,
    EscrowStatus,
    InvitePreviewResponse,
    AuthResult,
} from "../types/index";

export const escrowService = {
    async createEscrow(request: CreateEscrowRequest): Promise<EscrowResponse> {
        return api<EscrowResponse>("/escrows", {
            method: "POST",
            body: JSON.stringify(request),
        });
    },

    async getEscrow(id: string): Promise<EscrowResponse> {
        return api<EscrowResponse>(`/escrows/${id}`);
    },

    async listEscrows(status?: EscrowStatus): Promise<EscrowResponse[]> {
        const query = status ? `?status=${status}` : "";
        return api<EscrowResponse[]>(`/escrows${query}`);
    },

    async transitionStatus(
        id: string,
        toStatus: EscrowStatus,
        reason?: string,
    ): Promise<EscrowResponse> {
        return api<EscrowResponse>(`/escrows/${id}/status`, {
            method: "PATCH",
            body: JSON.stringify({ toStatus, reason }),
        });
    },

    async acceptInvite(inviteToken: string): Promise<AuthResult> {
        return api<AuthResult>("/escrows/invite/accept", {
            method: "POST",
            body: JSON.stringify({ inviteToken }),
            skipRefresh: true,
        });
    },

    async getInvitePreview(token: string): Promise<InvitePreviewResponse> {
        return api<InvitePreviewResponse>(
            `/escrows/invite/preview?token=${token}`,
            { skipRefresh: true },
        );
    },

    async deleteEscrow(id: string, password: string): Promise<void> {
        return api<void>(`/escrows/${id}`, {
            method: "DELETE",
            body: JSON.stringify({ password }),
        });
    },

    async initNewPayment(request: CreateEscrowRequest): Promise<any> {
        return api<any>("/payments/init-new", {
            method: "POST",
            body: JSON.stringify(request),
        });
    },

    async verifyAndCreate(data: {
        txnRef: string;
        amount: number;
        escrowData: CreateEscrowRequest;
    }): Promise<EscrowResponse> {
        return api<EscrowResponse>("/escrows/verify-and-create", {
            method: "POST",
            body: JSON.stringify(data),
        });
    },
};
